### ISSUE TITLE
CRS Extension Req18 error in TeamEngine with SOAP version 1.2

### ISSUE THREAD

**Issue by bangph** [opengeospatial/ets-wcs20#24]
It looks like the problem still here 
`https://lists.opengeospatial.org/pipermail/cite-forum/2011-August/000598.html`

and in this test case it send a SOAP request with CDATA, then it has problem

```
<ctl:soap-request version="1.2" charset="UTF-8">
                            <ctl:url>
                                <xsl:value-of select="$url" />
                            </ctl:url>
                            <ctl:method>post</ctl:method>
                            <ctl:header name="Content-type">text/xml</ctl:header>
                            <ctl:body><![CDATA[<?xml version="1.0" encoding="UTF-8"?>
								<wcs:GetCoverage xmlns:wcs="http://www.opengis.net/wcs/2.0"
									xmlns:gml="http://www.opengis.net/gml/3.2"
									xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:wcscrs="http://www.opengis.net/wcs/service-extension/crs/1.0"
									xsi:schemaLocation="http://www.opengis.net/wcs/2.0 ../../wcsAll.xsd"
									service="WCS" version="2.0.1">
									<wcs:CoverageId>]]><xsl:value-of select="$Capabilities//*[local-name()='CoverageSummary'][1]/*[local-name()='CoverageId']" /><![CDATA[</wcs:CoverageId>
						
 									   <wcs:DimensionTrim>
 									       <wcs:Dimension> ]]><xsl:value-of select="tokenize($DescribeCoverage//@axisLabels[1],' ')[1]" /><![CDATA[</wcs:Dimension>
 									       <wcs:TrimLow> ]]><xsl:value-of select="tokenize($DescribeCoverage//*[local-name()='lowerCorner'][1],' ')[1]" /><![CDATA[</wcs:TrimLow>
 									       <wcs:TrimHigh> ]]><xsl:value-of select="tokenize($DescribeCoverage//*[local-name()='upperCorner'][1],' ')[1]" /><![CDATA[</wcs:TrimHigh>
 									   </wcs:DimensionTrim>
 									   <wcs:DimensionTrim>
 									       <wcs:Dimension> ]]><xsl:value-of select="tokenize($DescribeCoverage//@axisLabels[1],' ')[2]" /><![CDATA[</wcs:Dimension>
 									       <wcs:TrimLow> ]]><xsl:value-of select="tokenize($DescribeCoverage//*[local-name()='lowerCorner'][1],' ')[2]" /><![CDATA[</wcs:TrimLow>
 									       <wcs:TrimHigh> ]]><xsl:value-of select="tokenize($DescribeCoverage//*[local-name()='upperCorner'][1],' ')[2]" /><![CDATA[</wcs:TrimHigh>
 									   </wcs:DimensionTrim> 
						
									<wcs:Extension>
									  <wcscrs:subsettingCrs>
									     ]]><xsl:value-of select="$Capabilities//*[local-name()='ServiceMetadata']//*[local-name()='CrsMetadata']/*[local-name()='crsSupported'][1]" /><![CDATA[
									  </wcscrs:subsettingCrs>
									  <wcscrs:outputCrs>
									    ]]><xsl:value-of select="$Capabilities//*[local-name()='ServiceMetadata']//*[local-name()='CrsMetadata']/*[local-name()='crsSupported'][1]" /><![CDATA[
									  </wcscrs:outputCrs>
									</wcs:Extension>
									
									</wcs:GetCoverage>]]></ctl:body>
                            <ctlp:XMLValidatingParser>
                                <ctlp:schemas>
                                    <ctlp:schema type="url">http://schemas.opengis.net/wcs/2.0/wcsAll.xsd</ctlp:schema>
                                </ctlp:schemas>
                            </ctlp:XMLValidatingParser>
                        </ctl:soap-request>
```

`Fatal Error: Error in call to extension function {public org.w3c.dom.NodeList com.occamlab.te.TECore.soap_request(org.w3c.dom.Document,java.lang.String) throws java.lang.Throwable}: DOMSender: no start node defined; SystemID:`

**Comment by bangph** [comment]
@jinsongdi123

**Comment by bangph** [comment]
@bermud  @keshav-nangare  Would you please have some suggestion ?

**Comment by dstenger** [comment]
@bangph Can you provide a service to reproduce the error? This helps us to verify a fix.

**Comment by rasdaman-doi** [comment]
@dstenger: I will prepare for you this endpoint in next days. It is good if CITE could tackle the problem more quickly, this question was asked one year ago and we have to wait too long.

**Comment by rasdaman-doi** [comment]
@dstenger: Here is the endpoint for your test http://moon.planetserver.eu:8084/rasdaman/ows?service=WCS&request=GetCapabilities&acceptversions=2.0.1

However, you need to do 2 things first:
+ Don't run CRS extension with CRS core as I showed a workaround here https://github.com/opengeospatial/ets-wcs20/issues/43#issuecomment-340423924 
+ And you could not reproduce the error on this issue from cite.opengeospatial.org/teamengine because of this issue https://github.com/opengeospatial/ets-wcs20/issues/22 (I even showed how to fix in the CTL script but no one answer). After the change for the CTL script in my local TeamEngine, then I had this problem for SOAP in CRS extension.

**Comment by keshav-nangare** [comment]
@dstenger 

I am able to reproduce this issue as suggested by rasdaman-doi.

If we added the CDATA in to the CTL- soap request and then the teamengine is trying to build the soap request from the CTL script at that time it checking the node-type.
If the node is not type of ELEMENT then the Dom object is null and we get the following error:

`Error in call to extension function {public org.w3c.dom.NodeList com.occamlab.te.TECore.soap_request(org.w3c.dom.Document,java.lang.String) throws java.lang.Throwable}: DOMSender: no start node defined`

If we removed CDATA from the CTL soap-request then its working fine.

If it is necessary to add CDATA in CTL soap-request then we need to update teamengine to support for this type of nodes.
Otherwise if we removed CDATA then there is no changes in teamengine.

**Comment by dstenger** [comment]
@keshav-nangare 

CDATA can be removed from CTL if this fixes the bug.

However, CDATA elements are used in several tests. Thus, it should be checked whether the other tests also fail. If this is not the case, why are the other tests not failing? Maybe, the bug is located in ```com.occamlab.te.TECore.soap_request()``` and parsers for other request encodings work?

**Comment by keshav-nangare** [comment]
We have two type of requests i.e. **`ctl:request`** and **`ctl:soap-request`**.

The main issue of the CDATA is with the ctl:soap-request, if CDATA is exist in the request <ctl:body> then all the body content is escaped while transforming the CTL test to executable style-sheet. Whenever we are trying to access this escaped body content it is getting as text node instead of element type. 

If the CDATA is not exist in the request body then it will keep this XML element/node as it is in the executable style-sheet.

For example:
# **&lt;ctl:request&gt; with CDATA:**

**CTL soap-request:**

```
<ctl:soap-request version="1.2" charset="UTF-8">
	<ctl:url>
		<xsl:value-of select="$url" />
	</ctl:url>
	<ctl:method>post</ctl:method>
	<ctl:header name="Content-type">text/xml</ctl:header>
	<ctl:body><![CDATA[<?xml version="1.0" encoding="UTF-8"?>
								<wcs:GetCoverage xmlns:wcs="http://www.opengis.net/wcs/2.0"
									xmlns:gml="http://www.opengis.net/gml/3.2"
									xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
									xsi:schemaLocation="http://www.opengis.net/wcs/2.0 ../../wcsAll.xsd"
									service="WCS" version="2.0.1">
									<wcs:CoverageId>]]><xsl:value-of
		select="$Capabilities//*[local-name()='CoverageSummary'][1]/*[local-name()='CoverageId']" /><![CDATA[</wcs:CoverageId>
						
 									   <wcs:DimensionTrim>
 									       <wcs:Dimension> ]]><xsl:value-of
		select="tokenize($DescribeCoverage//@axisLabels[1],' ')[1]" /> <![CDATA[</wcs:Dimension>
 									       <wcs:TrimLow> ]]>
		<xsl:value-of
			select="tokenize($DescribeCoverage//*[local-name()='lowerCorner'][1],' ')[1]" /><![CDATA[</wcs:TrimLow>
 									       <wcs:TrimHigh> ]]><xsl:value-of
		select="tokenize($DescribeCoverage//*[local-name()='upperCorner'][1],' ')[1]" /> <![CDATA[</wcs:TrimHigh>
 									   </wcs:DimensionTrim>
 									   <wcs:DimensionTrim>
 									       <wcs:Dimension> ]]><xsl:value-of
		select="tokenize($DescribeCoverage//@axisLabels[1],' ')[2]" /> <![CDATA[</wcs:Dimension>
 									       <wcs:TrimLow> ]]>
		<xsl:value-of
			select="tokenize($DescribeCoverage//*[local-name()='lowerCorner'][1],' ')[2]" /><![CDATA[</wcs:TrimLow>
 									       <wcs:TrimHigh> ]]><xsl:value-of
		select="tokenize($DescribeCoverage//*[local-name()='upperCorner'][1],' ')[2]" /> <![CDATA[</wcs:TrimHigh>
 									   </wcs:DimensionTrim> 
						
									<wcs:Extension>
									  <wcscrs:subsettingCrs>
									     ]]><xsl:value-of
		select="$Capabilities//*[local-name()='ServiceMetadata']//*[local-name()='CrsMetadata']/*[local-name()='crsSupported'][1]" /> <![CDATA[
									  </wcscrs:subsettingCrs>
									  <wcscrs:outputCrs>
									    ]]><xsl:value-of
		select="$Capabilities//*[local-name()='ServiceMetadata']//*[local-name()='CrsMetadata']/*[local-name()='crsSupported'][1]" /> <![CDATA[
									  </wcscrs:outputCrs>
									</wcs:Extension>
									
									</wcs:GetCoverage>]]>
	</ctl:body>

	<ctlp:XMLValidatingParser>
		<ctlp:schemas>
			<ctlp:schema type="url">http://schemas.opengis.net/wcs/2.0/wcsAll.xsd
			</ctlp:schema>
		</ctlp:schemas>
	</ctlp:XMLValidatingParser>

</ctl:soap-request>
```

**CTL soap-request after transformation:**

```
<ctl:request xmlns:ctl="http://www.occamlab.com/ctl">
	<ctl:url>
		<xsl:value-of select="$url" />
	</ctl:url>
	<ctl:method>post</ctl:method>
	<ctl:header name="Content-type">text/xml</ctl:header>
	<ctl:body>
		&lt;?xml version="1.0" encoding="UTF-8"?&gt;
		&lt;wcs:GetCoverage xmlns:wcs="http://www.opengis.net/wcs/2.0"
		xmlns:gml="http://www.opengis.net/gml/3.2"
		xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
		xsi:schemaLocation="http://www.opengis.net/wcs/2.0 ../../wcsAll.xsd"
		service="WCS" version="2.0.1"&gt;
		&lt;wcs:CoverageId&gt;
		<xsl:value-of
			te:loc="1909,file:/C:/TE_BASE/scripts/wcs/2.0.1/ctl/ext_crs.xml"
			select="$Capabilities//*[local-name()='CoverageSummary'][1]/*[local-name()='CoverageId']" />
		&lt;/wcs:CoverageId&gt;

		&lt;wcs:DimensionTrim&gt;
		&lt;wcs:Dimension&gt;
		<xsl:value-of
			te:loc="1912,file:/C:/TE_BASE/scripts/wcs/2.0.1/ctl/ext_crs.xml"
			select="tokenize($DescribeCoverage//@axisLabels[1],' ')[1]" />
		&lt;/wcs:Dimension&gt;
		&lt;wcs:TrimLow&gt;
		<xsl:value-of
			te:loc="1913,file:/C:/TE_BASE/scripts/wcs/2.0.1/ctl/ext_crs.xml"
			select="tokenize($DescribeCoverage//*[local-name()='lowerCorner'][1],' ')[1]" />
		&lt;/wcs:TrimLow&gt;
		&lt;wcs:TrimHigh&gt;
		<xsl:value-of
			te:loc="1914,file:/C:/TE_BASE/scripts/wcs/2.0.1/ctl/ext_crs.xml"
			select="tokenize($DescribeCoverage//*[local-name()='upperCorner'][1],' ')[1]" />
		&lt;/wcs:TrimHigh&gt;
		&lt;/wcs:DimensionTrim&gt;
		&lt;wcs:DimensionTrim&gt;
		&lt;wcs:Dimension&gt;
		<xsl:value-of
			te:loc="1917,file:/C:/TE_BASE/scripts/wcs/2.0.1/ctl/ext_crs.xml"
			select="tokenize($DescribeCoverage//@axisLabels[1],' ')[2]" />
		&lt;/wcs:Dimension&gt;
		&lt;wcs:TrimLow&gt;
		<xsl:value-of
			te:loc="1918,file:/C:/TE_BASE/scripts/wcs/2.0.1/ctl/ext_crs.xml"
			select="tokenize($DescribeCoverage//*[local-name()='lowerCorner'][1],' ')[2]" />
		&lt;/wcs:TrimLow&gt;
		&lt;wcs:TrimHigh&gt;
		<xsl:value-of
			te:loc="1919,file:/C:/TE_BASE/scripts/wcs/2.0.1/ctl/ext_crs.xml"
			select="tokenize($DescribeCoverage//*[local-name()='upperCorner'][1],' ')[2]" />
		&lt;/wcs:TrimHigh&gt;
		&lt;/wcs:DimensionTrim&gt;

		&lt;wcs:Extension&gt;
		&lt;wcscrs:subsettingCrs&gt;
		<xsl:value-of
			te:loc="1924,file:/C:/TE_BASE/scripts/wcs/2.0.1/ctl/ext_crs.xml"
			select="$Capabilities//*[local-name()='ServiceMetadata']//*[local-name()='CrsMetadata']/*[local-name()='crsSupported'][1]" />
		&lt;/wcscrs:subsettingCrs&gt;
		&lt;wcscrs:outputCrs&gt;
		<xsl:value-of
			te:loc="1927,file:/C:/TE_BASE/scripts/wcs/2.0.1/ctl/ext_crs.xml"
			select="$Capabilities//*[local-name()='ServiceMetadata']//*[local-name()='CrsMetadata']/*[local-name()='crsSupported'][1]" />
		&lt;/wcscrs:outputCrs&gt;
		&lt;/wcs:Extension&gt;

		&lt;/wcs:GetCoverage&gt;
	</ctl:body>
```	


# **&lt;ctl:request&gt; without CDATA:**

**CTL request:** 
```
<ctl:request>
   <ctl:url>
      <xsl:value-of select="$url" />
   </ctl:url>
   <ctl:method>post</ctl:method>
   <ctl:header name="Content-type">text/xml</ctl:header>
   <ctl:body>
      <csw:DescribeRecord xmlns:csw="http://www.opengis.net/cat/csw/2.0.2" service="CSW" 
version="2.0.2" schemaLanguage="http://www.w3.org/XML/Schema">
         <csw:TypeName>csw:Record</csw:TypeName>
      </csw:DescribeRecord>
   </ctl:body>
   <ctlp:XMLValidatingParser>
      <ctlp:schemas>
         <ctlp:schema type="url">http://schemas.opengis.net/wcs/2.0/wcsAll.xsd</ctlp:schema>
      </ctlp:schemas>
   </ctlp:XMLValidatingParser>
</ctl:request>

```

**CTL request after transformation:**
```
<ctl:request>
                  <ctl:url>
                     <xsl:value-of te:loc="113,file:/C:/TE_BASE/scripts/csw/2.0.2/ctl/main.xml"
                                   select="$csw.DescribeRecord.post.url"/>
                  </ctl:url>
                  <ctl:method>POST</ctl:method>
                  <ctl:body>
                     <csw:DescribeRecord service="CSW" version="2.0.2" schemaLanguage="http://www.w3.org/XML/Schema">
                        <csw:TypeName>csw:Record</csw:TypeName>
                     </csw:DescribeRecord>
                  </ctl:body>
                  <p:XMLValidatingParser.CSW/>
               </ctl:request>
```




While building the request the teamengine is checking the node type and according to type it is doing further processing.


In **`ctl:request`**:

This type of escaped content is working in `ctl:request` because if the teamengine found the body content as text then it will get all the text-content as part request body and then request is invoked.

In **`ctl:soap-request`**:

The soap-request is not supporting this type of content, if the the teamengine found the node type as text(not as element)then it return the null byte array that's why its failing with the above exception.

This is the reason other tests are working but failing with soap request.

**Comment by keshav-nangare** [comment]
@dstenger 

There is one more solution for this issue i.e. we can convert this text node into element and then we can do process in soap request if it is best approach?

**Comment by dstenger** [comment]
I propose to adjust the implementation of ```soap-requqest``` being able to process elements in CDATA.

**Comment by dstenger** [comment]
Will be fixed by https://github.com/opengeospatial/teamengine/issues/276.

As soon as fix will have been done and testing will have been successful, this issue can be closed.

**Comment by dstenger** [comment]
After fix opengeospatial/teamengine#276 has been done, the request cannot be parsed as it is erroneous. Prefix wcscrs is not bound to a namespace. Thus, the request must also be fixed.

Solution:
Namespace binding must be added to the request:
```
xmlns:wcscrs="http://www.opengis.net/wcs/crs/1.0"
```

This also applies to ```wcs2:crs-req17```. This request body should also be fixed.

**Comment by keshav-nangare** [comment]
Added missing namespace in request body.

Updated with this commit: c3889611e98c8665841178218510cefdf4945385
